### DISCUSSION TITLE
Question: API TS 类型定义 有无计划完善？

### DISCUSSION THREAD

**Discussion by x011223** [antvis/G2#5465] [Q&A] [answered]
最新版本(5.0.22)，官网明明白白写着的方法，调用就是报错。
![image](https://github.com/antvis/G2/assets/31724313/1e96b2a9-7264-46ce-9199-bb3c6af3778f)
![image](https://github.com/antvis/G2/assets/31724313/6d4f8e7d-4c61-422d-bf55-07a01744941b)

除此之外，充斥着大量 object 类型。

![image](https://github.com/antvis/G2/assets/31724313/e56d83af-3e79-439a-a9ea-7134d67ed344)

``` json
 "@antv/g2": "^5.0.22",
 ```

**Comment by pearmini** [comment]
截图，让我看看 `chart` 对象的创建过程？

**Reply by x011223** [reply]
下面是实例创建完整代码 以及 截图
 ```json
 "typescript": "latest",
"react": "^18.2.0",
"react-dom": "^18.2.0",
"@antv/g2": "^5.0.22"
 ```
 ```typescript
 useEffect(() => {
    const chart = new Chart({
      width: 500,
      height: 400,
      container,
      // theme: 'classicDark',
      theme: 'classic',
      type: 'view',
      autoFit: true,
      axis: {
        y: { tickCount: 5 },
      },
      tooltip: {
        title: 'name', // 标题
        items: ['type'], // 数据项
        showTitle: false,
        showMarkers: false,
        crosshairs: {
          type: 'line',
        },
      },
    });
    chart
      .interval()
      // 绑定数据
      .data(data)
      .transform({ type: 'stackY' })
      // .transform({ type: 'sortX', by: 'y', reverse: true })
      .encode('x', 'x') // 编码 x 通道
      .encode('y', 'y') // 编码 y 通道
      .animate('enter', { type: 'scaleInY', duration: 800 })
      .animate('exit', { type: 'scaleOutY', duration: 800 });

    // 堆叠
    if (barType === BarType.double_stack) {
      chart.encode('color', 'type');
    }

    chart.render().then().catch((error) => {
      console.log(`${container} render error: ${error}`);
    });

    return () => {
      chart.clear();
      chart.destroy();
      chartRef.current = null;
    };
  }, [width, height, container, data, barType]);
  ```

![image](https://github.com/antvis/G2/assets/31724313/ae018153-7b3d-4cee-bf6e-4e18ef815201)

**Reply by pearmini** [reply]
object 是因为 `encode` 这些方法是通过映射类型生成的：https://github.com/antvis/G2/blob/v5/src/api/types.ts#L92

**Reply by pearmini** [reply]
是 `chart.clear` 又提示，但是 `chart.destroy` 没有提示？

**Reply by x011223** [reply]
是的，有没有可能是ts版本 比较高 解释器 逻辑不一样 导致的？映射类型是出于什么考虑啊，这样的话代码智能提示完全没有，文档 又不全，做一个 tooltip / legned 同时满足需求的 柱形图都感觉到深深的痛苦了。。。

**Reply by pearmini** [reply]
- 映射类型是把底层的 spec 类型映射为 api 类型，可以看看[这里](https://g2.antv.antgroup.com/manual/core/api)。
- tooltip 和 legend 是遇见什么问题了？

**Reply by x011223** [reply]
``` ts
// 如何 替换y轴 title的值 例如 改成 单位: 元
  chart.axis('y', {
    style: {
      titlePosition: 't',
    },
  });
// 这么写  legend 在画布中布局无效。
chart.legend('type', {
  layout: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
});

// tool 同类型的值在 tooltip怎么同时显示？
const chart = new Chart({
  width,
  height,
  container,
  theme: 'classic',
  // type: 'view',
  autoFit: true,
  tooltip: {
    items: [
      { channel: 'y' },
    ],
  },
});
```
这几个东西在v4 都是写过的，但是现在要么API 不再支持要么文档没有明写支不支持 但是传进去无效。

![image](https://github.com/antvis/G2/assets/31724313/99d37f40-6459-468a-b267-0e82f6744abd)
![image](https://github.com/antvis/G2/assets/31724313/de3329a6-4e29-4e26-916f-e8070b97f7d7)

**Reply by pearmini** [reply, marked as answer]
- 问题一：legend 是和通道绑定，而不是字段，`mark.legend('color', {}}` 而不是 `mark.legend('type', {}`
- 问题二：`mark.axis('y', { title: 'hello' })`
- 问题三：`chart.interaction("tooltip", { shared: true })`

**Comment by ghost** [comment]
尝试着来了解了一下G2，但是简单的使用下来简直不要太痛苦，如图中有很多这种Object类型，但是文档中找不到相关类型信息说明，本想着打开TS类型定义文件，正如标题中的问题，充斥着大量的Object，完全用不明白。
<img width="1263" alt="image" src="https://github.com/antvis/G2/assets/103547464/ef15e65b-5d1b-4c4a-8e75-0ec1b35c0949">


我认为要么类型定义清楚，能与文档中的对应上也可以，要么在文档中说明，否则只能说使用体验极差，完全靠猜，还没有思路。
